### PR TITLE
Earth/Ting-Yi/Taylor

### PR DESCRIPTION
**Pull request by TaylorMililani**
# Assignment Submission: Rideshare Rails

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These questions should be answered by all team members together, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Describe the types of entity relationships you set up in your project and why you set up the relationships that way | We set up trip belongs to both passenger and driver, and passenger and driver both have many trips.
Describe the role of model validations in your application | Prevents the creation of invalid data.
How did your team break up the work to be done? | We split up Passenger and Driver tasks, and did most Trip tasks together. 
What features did you choose to prioritize in your project, and what features, if any, did you have to set aside to meet the deadline? | We prioritized required functionality and also the complete trip feature, while we set css to the back burner.
What was one thing that your team collectively gained more clarity on after completing this assignment? | Relationships between models, model tests, and nested routes.
What are two discussion points that you and your team discussed when giving/receiving feedback from each other that you would be willing to share? | We talked about how we both prefer direct communication.
Optional: What is the URL of your deployed Heroku app? | https://desolate-atoll-20205.herokuapp.com/

### CODE REVIEW

File: app/controllers/drivers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @driver = Driver.new
+  end
+
+  def create
+    @driver = Driver.new(driver_params) 
+    if @driver.save 
+      redirect_to driver_path(id: @driver[:id]) 
+      return
+    else 
+      render :new, status: :bad_request
```
Comment (CheezItMan) [inline app/controllers/drivers_controller.rb:24]: 👍 Nice!

File: app/controllers/drivers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  def edit
+    @driver = Driver.find_by(id: params[:id])
+
+    if @driver.nil? 
+      redirect_to drivers_path 
+      return
+    elsif !(@driver.isactive)
+      render :file => "#{Rails.root}/public/404.html",  layout: false, status: :not_found
```
Comment (CheezItMan) [inline app/controllers/drivers_controller.rb:36]: 😁 Well done

File: app/models/driver.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    sum = 0
+    every_costs.each do |trip|
+      if trip.cost <= 1.65 * 100
+        sum += trip.cost * 0.8
+      else
+        sum += (trip.cost - 1.65 * 100) * 0.8
+      end
+    end
+    return (sum / 100).round(2)
+  end
```
Comment (CheezItMan) [inline app/models/driver.rb:20]: Just a note that you can use enumerables as well.

```suggestion
  sum = every_costs.sum do |trip|
    if trip.cost.nil?
      0
    elsif trip.cost < 1.65 * 100
      trip.cost * 0.8
    else
      (trip.cost - 1.65 * 100) * 0.8 
  end
```

File: app/models/driver.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  def ave_rating
+    rating_by_passengers = self.trips.where.not(rating: nil)
+    return nil if rating_by_passengers.empty?
+
+    sum = rating_by_passengers.sum { |trip| trip.rating}
+    return (sum / rating_by_passengers.count.to_f).round(1)
+  end
```
Comment (CheezItMan) [inline app/models/driver.rb:28]: 👍

File: app/models/passenger.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class Passenger < ApplicationRecord
+  has_many :trips
+
+  validates :name, presence: true
+  validates :phone_num, presence: true
+
+  def total_spent
```
Comment (CheezItMan) [inline app/models/passenger.rb:7]: Just a note that you can do an enumerable here as well.

File: app/models/trip.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Trip < ApplicationRecord
+  belongs_to :driver
+  belongs_to :passenger
+
+  validates :rating, :inclusion => { :in => [1,2,3,4,5,nil] }
```
Comment (CheezItMan) [inline app/models/trip.rb:5]: 👍

File: app/views/passengers/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    <ul class="errors">
+      <% @passenger.errors.each do |column, message| %>
+        <li>
+          <strong><%= column.capitalize %></strong> <%= message %>
+        </li>
+      <% end %>
+    </ul>
+    <% end %>
+  </article>
+</section>
```
Comment (CheezItMan) [inline app/views/passengers/_form.html.erb:39]: Just a note, putting the validations a the bottom will make it easier for a user to miss.

File: test/models/driver_test.rb [hunk 8]
```diff
@@ -47,35 +47,52 @@
...
 
-    describe "can go offline" do
-      # Your code here
+        assert_nil(no_trip_driver.total_earnings)
+      end
     end
 
-    # You may have additional methods to test
+    describe "ave_rating" do
+      it "calculates the average rating of a driver based on all trips" do
```
Comment (CheezItMan) [inline test/models/driver_test.rb:81]: I would also have a test with some complete trips and at least 1 incomplete trips (nil rating).

File: test/models/trip_test.rb [hunk 9]
```diff
@@ -1,24 +1,49 @@
...
-    # Your code here
+    new_trip.save
+    trip = Trip.first
+    [:driver_id, :passenger_id, :date, :rating, :cost].each do |field|
+      # Assert
+      expect(trip).must_respond_to field
+    end
   end
 
   describe "relationships" do
```
Comment (CheezItMan) [inline test/models/trip_test.rb:27]: I would also include a test that you can't create a trip without a passenger and driver.

File: test/controllers/drivers_controller_test.rb [hunk 10]
```diff
@@ -2,174 +2,379 @@
...
 
   describe "index" do
     it "responds with success when there are many drivers saved" do
       # Arrange
       # Ensure that there is at least one Driver saved
+      driver = @driver
 
       # Act
-
+      get "/drivers"
```
Comment (CheezItMan) [inline test/controllers/drivers_controller_test.rb:16]: Better to use the path method

```suggestion
      get drivers_path
```

File: test/controllers/drivers_controller_test.rb [hunk 11]
```diff
@@ -2,174 +2,379 @@
...
     end
 
     it "responds with success when there are no drivers saved" do
       # Arrange
+      Driver.delete_all
+      
       # Ensure that there are zero drivers saved
-
       # Act
+      get "/drivers"
```
Comment (CheezItMan) [inline test/controllers/drivers_controller_test.rb:29]:
```suggestion
      get drivers_path
```

File: test/controllers/drivers_controller_test.rb [hunk 12]
```diff
@@ -2,174 +2,379 @@
...
+      get new_driver_path
+
+      must_respond_with :success
+    end
+
     it "responds with success" do
+      get new_driver_path(Driver.first.id)
+
+      must_respond_with :success
     end
```
Comment (CheezItMan) [inline test/controllers/drivers_controller_test.rb:87]: I don't understand why this is here.  You don't create a new driver with another driver's id.

File: test/controllers/passengers_controller_test.rb [hunk 13]
```diff
@@ -1,31 +1,327 @@
...
+    @passenger = Passenger.create(name: "Judy", phone_num: "360-555-0987")
+  end
   describe "index" do
-    # Your tests go here
+    it "responds with success when there are many passengers saved" do
+      # Arrange
+      passenger = @passenger
+
+      # Act
+      get "/passengers"
```
Comment (CheezItMan) [inline test/controllers/passengers_controller_test.rb:13]:
```suggestion
      get passengers_path
```

File: test/controllers/trips_controller_test.rb [hunk 14]
```diff
@@ -1,23 +1,266 @@
...
+    let (:invalid_trip) {
+      {
+          trip: {
+              date: "2016-04-05",
+              rating: 50,
+              cost: 12.0
+          }
+      }
+    }
+    it "can update an existing trip with valid information accurately, and redirect" do
```
Comment (CheezItMan) [inline test/controllers/trips_controller_test.rb:144]: I would also include a test that the driver's average rating changes appropriately.

File: test/controllers/trips_controller_test.rb [hunk 15]
```diff
@@ -1,23 +1,266 @@
...
+      
+      expect(@driver.isactive).must_equal false
+      expect(@passenger.isactive).must_equal true
+
+      # Assert
+      must_redirect_to trip_path(@trip.id)
+    end
+  end
+
+  describe "complete_trip" do
```
Comment (CheezItMan) [inline test/controllers/trips_controller_test.rb:250]: I would also include a test to verify that the driver's average rating changed appropriately.

File: config/routes.rb [hunk 16]
```diff
@@ -0,0 +1,21 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+  root 'homepages#index'
+  resources :drivers do
+    resources :trips, only: [:index]
+  end
+
+  resources :passengers do
+    resources :trips, only: [:index, :new, :create]
```
Comment (CheezItMan) [inline config/routes.rb:10]: Do you really need the new trip form?  You can just automatically create a trip from the link.

File: config/routes.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+  root 'homepages#index'
+  resources :drivers do
+    resources :trips, only: [:index]
```
Comment (CheezItMan) [inline config/routes.rb:6]: Do you use this nested route?

File: config/routes.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  resources :passengers do
+    resources :trips, only: [:index, :new, :create]
+  end
+
+  resources :trips, except: [:index] do
+    resources :drivers, only: [:index]
+    resources :passengers, only: [:index]
+  end
```
Comment (CheezItMan) [inline config/routes.rb:16]: You really don't need all these routes.  Think about this a bit.

File: app/controllers/trips_controller.rb [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+  def destroy
+    @trip = Trip.find_by(id: params[:id])
+    if @trip.nil?
+      render :file => "#{Rails.root}/public/404.html",  layout: false, status: :not_found
+      return
+    end
+
+    driver = @trip.driver
+    passenger = @trip.passenger
+    if !(driver.isactive) && !(passenger.isactive)
```
Comment (CheezItMan) [inline app/controllers/trips_controller.rb:95]: I'm unclear why you couldn't delete a trip if the passenger or driver was active?

**Review by CheezItMan — COMMENTED** [review]
# Rideshare Rails



## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Practices collaborating with git, and all team members contribute git commits and best git practices | ✔️
| Demonstrates understanding of relationships by giving accurate answers to the reflection questions and appropriate code in the models | ✔️
| Logic to calculate a driver's average rating and total earnings is located in the model, and it has unit tests | ✔️
| There are reasonable tests to test the validation requirements for all models | ✔️
| There are reasonable tests to test the relationship requirements for all models | ✔️
| There are reasonable tests to test the controller actions for all controllers | ✔️
| The app has an attractive and usable user interface | ✔️, really nice looking!
| The app uses/is compatible with database seeds | ✔️
| Router code is clean: uses `resources` and RESTful routes | ⚠️ Oops, you've got some extra routes defined in your resources.  See my inline comments. 

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| On the passenger's details page, I want to be able to see total charged, list of trips, a link to edit, and delete | ✔️
| When adding a new passenger, I want to see errors and validations that show that a passenger must be provided a name and a phone number, so that I cannot make a passenger without name or phone number | ✔️
| On the passenger's details page, I can create a new trip for this passenger, with an assigned driver and no rating | ✔️
| On the driver's details page, I want to be able to see total earnings, average rating, list of trips, a link to edit, and delete | ✔️
| When adding a new driver, I want to see errors and validations that show that a driver must be provided a name and VIN, so that I cannot make a driver without name or VIN | ✔️
| On the trip's detail page, I want to be able to view details, assign a rating, navigate to the trip's passenger, driver, a link to edit, and delete | ✔️, except that I can't delete unless the driver and passenger are already deleted... odd.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 5+ in Functional Requirements | 💚





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

Nice work Tiny-Yi & Taylor.  This is truly outstanding.  Take a look at my comments and let me know what questions you have.

File: app/controllers/trips_controller.rb (hunk 19 above)
Comment (ichbinorange) [reply app/controllers/trips_controller.rb:95]: If the trip is deletable when either driver or passenger is still active, then either of them couldn't see the trip after it's been deleted when they want to see it. If I was the user, I still want to see my trip, that was what I thought back to that moment.  Does it make sense?
